rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,987 @@
+6. Applies the resulting assignment to the `IRModule` using the new `backend` field on every expression.
+7. An evolutionary search algorithm may optionally run after the above, and will attempt to replace ‘op’ kernels
+   (which use a library) with ‘graph’ kernels (arbtrary sub-graph), but only if there’s a unique graph backend enabled.
+8. An intercept
+   ([here](https://github.com/mbs-octoml/mbs-tvm/blob/52d8780e879a9115b8a93e505bcd3a6c2646c61f/src/relay/transforms/fuse_ops.cc#L1402)
+   and
+   [here](https://github.com/mbs-octoml/mbs-tvm/blob/52d8780e879a9115b8a93e505bcd3a6c2646c61f/python/collage/optimizer/_optimizer.py#L48))
+   in `fuse_ops.cc` to actually effect the fusion for BYOC backends depending on the new `backend` field.
+9. An intercept
+   ([here](https://github.com/mbs-octoml/mbs-tvm/blob/52d8780e879a9115b8a93e505bcd3a6c2646c61f/src/relay/backend/te_compiler_cache.cc#L284)
+   and
+   [here](https://github.com/mbs-octoml/mbs-tvm/blob/52d8780e879a9115b8a93e505bcd3a6c2646c61f/python/collage/backend/collage_strategies.py#L18))
+   in `te_compiler.cc` to take over the selection of `OpStrategy` based on the `backend` field.
+
+Note that the 'v1' prototype only supports `IRModules` with a single `"main"` whose body is in the ‘pure dataflow’ Relay
+subset. Ie only calls, tuples, tuple projections, function variables and constants are supported.
+
+In comparison to the 'v1' prototype, this design:
+
+1. Shifts the search to occur at the very start of the compilation flow rather than just before `FuseOps` so that
+   Collage sees the same input model as would an existing `partition_for_<toolchain>` BYOC function. This change allows
+   us to use the existing BYOC patterns/predicates to guide the selection of candidate partitions instead of requiring
+   new patterns to be added to support the combination of models and BYOC toolchains. It also ensures the existing BYOC
+   lowering functions see partitions before any TVM-lowering specific passes have been applied, such as
+   `qnn::transform::Legalize` and `transform::CombineParallelConv2D`.
+2. Builds on the existing support for heterogeneous `Target`s to represent the menu of available toolchains to use
+   during search. In particular, we want to allow users to blend `on_device` annotations (to express preferences for
+   which devices should execute which sub-graphs) with Collage (to find the best partitions respecting those device
+   preferences).
+3. Uses the existing convention for `"Primitive"`, `"Composite"` and `"Compiler"` attributes on Relay `Function`s to
+   encode partitioning choices.
+4. Does not treat library integrations (eg for `CuDnn`) differently from toolchain integrations (eg for `TensorRT`). See
+   Appendix B for a sketch of the issues.
+5. Supports all of Relay.
+6. Is implemented almost entirely in C++.
+
+However:
+
+6. We have only re-implemented the 'op-level' dynamic-programming based search strategy from the paper. Though the paper
```

Comment:
It might be worth clarifying that here op-level will consider sub-graphs of operators as well. Having read the paper, I can understand what is meant here, however, it has the tendency to confuse the reader to assume the current implementation only works single relay op-level, which I think is not true.

Or Am I also confused :p ?